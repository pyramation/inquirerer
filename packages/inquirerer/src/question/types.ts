export interface Value {
  name: string;
  value: boolean;
}
export interface OptionValue {
  name: string;
  value: any;
}

export interface Validation {
  type?: string;
  success: boolean;
  reason?: string;
}

export interface BaseQuestion {
    name: string;
    type: string;
    default?: any;
    useDefault?: boolean;
    required?: boolean;
    message?: string;
    description?: string;
    validate?: (input: any, obj: any) => Validation | boolean;
    sanitize?: (input: any, obj: any) => any;
    pattern?: string;
  }
  
  export interface ConfirmQuestion extends BaseQuestion {
    type: 'confirm';
    default?: boolean;  // Defaults are typically boolean for confirm types
  }
  
  export interface AutocompleteQuestion extends BaseQuestion {
    type: 'autocomplete';
    options: (string|OptionValue)[];
    maxDisplayLines?: number;
    returnFullResults?: boolean;
    allowCustomOptions?: boolean;
  }
  
  export interface CheckboxQuestion extends BaseQuestion {
    type: 'checkbox';
    options: (string|OptionValue)[];
    maxDisplayLines?: number;
    returnFullResults?: boolean;
    default?: string[];
  }
  
  export interface TextQuestion extends BaseQuestion {
    type: 'text';
    default?: string;
  }

  export interface NumberQuestion extends BaseQuestion {
    type: 'number';
    default?: number;
  }
  
  export type Question = ConfirmQuestion | AutocompleteQuestion | CheckboxQuestion | TextQuestion | NumberQuestion;