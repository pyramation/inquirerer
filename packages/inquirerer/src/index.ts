export * from './prompt';
export * from './question';