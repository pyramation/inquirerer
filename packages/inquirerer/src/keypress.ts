type KeyHandler = () => void;

export const KEY_CODES = {
  UP_ARROW: '\u001b[A',
  DOWN_ARROW: '\u001b[B',
  RIGHT_ARROW: '\u001b[C',
  LEFT_ARROW: '\u001b[D',
  ENTER: '\r',
  SPACE: ' ',
  CTRL_C: '\u0003',
  BACKSPACE: '\x7f',  // Commonly used BACKSPACE key in Unix-like systems
  BACKSPACE_LEGACY: '\x08'  // For compatibility with some systems
};

export class TerminalKeypress {
  private listeners: Record<string, KeyHandler[]> = {};
  private active: boolean = true;


  constructor() {
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.setEncoding('utf8');
    this.setupListeners();
  }


  private setupListeners(): void {
    process.stdin.on('data', (key: string) => {
      if (!this.active) return;
      const handlers = this.listeners[key];
      handlers?.forEach(handler => handler());
      if (key === KEY_CODES.CTRL_C) { // Ctrl+C
        process.exit();
      }
    });
  }

  on(key: string, callback: KeyHandler): void {
    if (!this.listeners[key]) {
      this.listeners[key] = [];
    }
    this.listeners[key].push(callback);
  }

  off(key: string, callback: KeyHandler): void {
    if (this.listeners[key]) {
      const index = this.listeners[key].indexOf(callback);
      if (index !== -1) {
        this.listeners[key].splice(index, 1);
      }
    }
  }

  pause(): void {
    this.active = false;
  }

  resume(): void {
    this.active = true;
  }

  destroy(): void {
    process.stdin.setRawMode(false);
    process.stdin.pause();
    process.stdin.removeAllListeners('data');
  }
}
