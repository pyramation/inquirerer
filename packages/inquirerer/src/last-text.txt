{
  "question": {
    "type": "text",
    "name": "last",
    "required": true,
    "message": "Enter your last name"
  },
  "obj": {
    "fruitSearch": "Apricot",
    "first": "adsf",
    "last": "asdf"
  }
}