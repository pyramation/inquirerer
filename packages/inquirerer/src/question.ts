export interface Question {
  name: string;
  type?: string; // This can be used for further customizations like validating input based on type

  options?: string[],

  maxDisplayLines?: number;  // Optional parameter to limit the number of visible options

  returnFullResults?: boolean;
  allowCustomOptions?: boolean;

  required?: boolean;
}
