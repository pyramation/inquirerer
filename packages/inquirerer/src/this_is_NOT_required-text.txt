{
  "question": {
    "name": "this_is_NOT_required",
    "type": "checkbox",
    "maxDisplayLines": 5,
    "options": [
      "Apple",
      "Apricot",
      "Avocado",
      "Banana",
      "Blackberry",
      "Blueberry",
      "Boysenberry",
      "Cherry",
      "Clementine",
      "Coconut",
      "Cranberry",
      "Date",
      "Durian",
      "Elderberry",
      "Fig",
      "Grape",
      "Grapefruit",
      "Guava",
      "Honeydew",
      "Kiwi",
      "Kumquat",
      "Lemon",
      "Lime",
      "Lychee",
      "Mango",
      "Melon",
      "Mulberry",
      "Nectarine",
      "Orange",
      "Papaya",
      "Peach",
      "Pear",
      "Persimmon",
      "Pineapple",
      "Plum",
      "Pomegranate",
      "Pomelo",
      "Quince",
      "Raspberry",
      "Redcurrant",
      "Strawberry",
      "Starfruit",
      "Tangerine",
      "Ugli Fruit",
      "Vanilla",
      "Watermelon",
      "Xigua (Chinese Watermelon)",
      "Yellow Plum",
      "Zucchini"
    ],
    "required": true
  },
  "obj": {
    "fruitSearch": "Apple",
    "first": null,
    "last": "sdf",
    "autocomp": "Apple",
    "this_is_NOT_required": []
  }
}