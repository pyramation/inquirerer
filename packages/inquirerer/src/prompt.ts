import chalk from 'chalk';
import readline from 'readline';
import { Readable, Writable } from 'stream';

import { KEY_CODES, TerminalKeypress } from './keypress';
import { AutocompleteQuestion, CheckboxQuestion, ConfirmQuestion, OptionValue, Question, TextQuestion, Validation, Value } from './question';

interface ManPageInfo {
  commandName: string;
  questions: Question[];
  author?: string;
  description?: string;
}

const validationMessage = (question: Question, ctx: PromptContext): string => {
  if (ctx.numTries === 0 || ctx.validation.success) {
    return ''; // No message if first attempt or validation was successful
  }

  if (ctx.validation.reason) {
    return chalk.red(`The field "${question.name}" is invalid: ${ctx.validation.reason}\n`);
  }

  switch (ctx.validation.type) {
    case 'required':
      return chalk.red(`The field "${question.name}" is required. Please provide a value.\n`);
    case 'pattern':
      return chalk.red(`The field "${question.name}" does not match the pattern: ${question.pattern}.\n`);
    default:
      return chalk.red(`The field "${question.name}" is invalid. Please try again.\n`);
  }

  return ''; // Return empty string if no specific conditions are met
};
class PromptContext {
  numTries: number = 0;
  needsInput: boolean = true;
  validation: Validation = { success: false };

  constructor() { }

  tryAgain(validation: Partial<Validation>): void {
    this.numTries++;
    this.needsInput = true;
    this.validation = { ...this.validation, ...validation, success: false };
  }

  nextQuestion(): void {
    this.numTries = 0;
    this.needsInput = false;
    this.validation = { success: true };
  }

  process(validation: Validation | boolean): Validation {
    if (typeof validation === 'boolean') {
      if (validation) {
        this.nextQuestion();
      } else {
        this.tryAgain({ type: 'validation' });
      }
    } else {
      if (validation.success) {
        this.nextQuestion();
      } else {
        this.tryAgain(validation);
      }
    }
    return this.validation;
  }
}


function generatePromptMessage(question: Question, ctx: PromptContext): string {
  let promptMessage: string = '';
  if (question.message) {
    promptMessage = chalk.whiteBright(question.message) + '\n';
  }
  promptMessage += `${chalk.white('[')}${chalk.green('--' + question.name)}${chalk.white(']:')}\n`;

  promptMessage = validationMessage(question, ctx) + promptMessage;

  switch (question.type) {
    case 'confirm':
      promptMessage += `(y/n)${question.default !== undefined ? ` [${question.default ? 'y' : 'n'}]` : ''}`;
      break;
    case 'text':
      if (question.default) {
        promptMessage += ` [${question.default}]`;
      }
      break;
    case 'autocomplete':
    case 'checkbox':
      // For these types, you might want to show the default selected options if any
      if (question.options && question.default) {
        const defaultOptions = Array.isArray(question.default) ? question.default : [question.default];
        const defaultText = defaultOptions.join(', ');
        promptMessage += ` [${defaultText}]`;
      }
      break;
  }

  return promptMessage;
}

export interface InquirererOptions {
  noTty?: boolean;
  input?: Readable;
  output?: Writable;
  useDefaults?: boolean;
  globalMaxLines?: number;
}
export class Inquirerer {
  private rl: readline.Interface | null;
  private keypress: TerminalKeypress;
  private noTty: boolean;
  private output: Writable;
  private input: Readable;
  private useDefaults: boolean;
  private globalMaxLines: number;

  constructor(
    options?: InquirererOptions
  ) {
    const {
      noTty = false,
      input = process.stdin,
      output = process.stdout,
      useDefaults = false,
      globalMaxLines = 10
    } = options ?? {}

    this.useDefaults = useDefaults;
    this.noTty = noTty;
    this.output = output;
    this.input = input;
    this.globalMaxLines = globalMaxLines;

    if (!noTty) {
      this.rl = readline.createInterface({
        input,
        // dissallow readline from prompting user, since we'll handle it!
        output
        // : new Writable({
        //   write(chunk, encoding, callback) {
        //     callback(); // Do nothing with the data
        //   }
        // })
      });
      this.keypress = new TerminalKeypress(noTty, input);
    } else {
      this.rl = null;
    }
  }

  clearScreen() {
    // same as console.clear()
    this.output.write('\x1Bc'); // This is the escape sequence to clear the terminal screen.
  }

  write(message: string) {
    this.output.write(message);
  }

  log(message: string) {
    this.output.write(message + '\n');
  }

  getInput(input: string) {
    return `${chalk.white.bold('$')} ${input}`;
  }

  getPrompt(question: Question, ctx: PromptContext, input: string) {
    const promptMessage = generatePromptMessage(question, ctx);
    return promptMessage + this.getInput(input);
  }
  displayPrompt(question: Question, ctx: PromptContext, input: string) {
    const prompt = this.getPrompt(question, ctx, input);
    this.log(prompt);
  }

  generateManPage(opts: ManPageInfo): string {

    let manPage = `NAME\n\t${opts.commandName} ${opts.description ?? ''}\n\n`;

    // Constructing the SYNOPSIS section with required and optional arguments
    let requiredArgs = '';
    let optionalArgs = '';

    opts.questions.forEach(question => {
      if (question.required) {
        requiredArgs += ` --${question.name} <${question.name}>`;
      } else {
        optionalArgs += ` [--${question.name}${question.default ? `=${question.default}` : ''}]`;
      }
    });

    manPage += `SYNOPSIS\n\t${opts.commandName}${requiredArgs}${optionalArgs}\n\n`;
    manPage += `DESCRIPTION\n\tUse this command to interact with the application. It supports the following options:\n\n`;

    opts.questions.forEach(question => {
      manPage += `${question.name.toUpperCase()}\n`;
      manPage += `\tType: ${question.type}\n`;
      if (question.message) {
        manPage += `\tSummary: ${question.message}\n`;
      }
      if (question.description) {
        manPage += `\tDescription: ${question.description}\n`;
      }
      if ('options' in question) {
        const optionsList = Array.isArray(question.options)
          ? question.options.map(opt => typeof opt === 'string' ? opt : `${opt.name} (${opt.value})`).join(', ')
          : '';
        manPage += `\tOptions: ${optionsList}\n`;
      }
      if (question.default !== undefined) {
        manPage += `\tDefault: ${JSON.stringify(question.default)}\n`;
      }
      if (question.required) {
        manPage += `\tRequired: Yes\n`;
      } else {
        manPage += `\tRequired: No\n`;
      }
      manPage += '\n';
    });

    manPage += `EXAMPLES\n\tExample usage of \`${opts.commandName}\`.\n\t$ ${opts.commandName}${requiredArgs}${optionalArgs}\n\n`;
    manPage += opts.author ? `AUTHOR\n\t${opts.author}\n` : '';
    return manPage;
  }

  private isValidatableAnswer(answer: any): boolean {
    return answer !== undefined;
  }

  private validateAnswer(question: Question, answer: any, obj: any, ctx: PromptContext): Validation {
    const validation = this.validateAnswerPattern(question, answer);
    if (!validation.success) {
      return ctx.process(validation);
    }

    if (question.validate) {
      const customValidation = question.validate(answer, obj);
      return ctx.process(customValidation);
    }

    return ctx.process({
      success: true
    });
  }

  private isValid(question: Question, obj: any, ctx: PromptContext): boolean {
    if (this.isValidatableAnswer(obj[question.name])) {
      obj[question.name] = this.sanitizeAnswer(question, obj[question.name], obj);
      const validationResult = this.validateAnswer(question, obj[question.name], obj, ctx);
      if (!validationResult.success) {
        return false;
      }
    }

    if (question.required && this.isEmptyAnswer(obj[question.name])) {
      ctx.tryAgain({ type: 'required' });
      return false;
    }

    return true;
  }

  private validateAnswerPattern(question: Question, answer: any): Validation {
    if (question.pattern && typeof answer === 'string') {
      const regex = new RegExp(question.pattern);
      const success = regex.test(answer);
      if (success) {
        return {
          success
        }
      } else {
        return {
          type: 'pattern',
          success: false,
          reason: question.pattern
        }
      }
    }
    return {
      success: true
    }
  }

  private isEmptyAnswer(answer: any): boolean {
    switch (true) {
      case answer === undefined:
      case answer === null:
      case answer === '':
      case Array.isArray(answer) && answer.length === 0:
        return true;
    }
    return false;
  }

  private sanitizeAnswer(question: Question, answer: any, obj: any): any {
    if (question.sanitize) {
      return question.sanitize(answer, obj);
    }
    return answer;
  }


  public async prompt<T extends object>(
    argv: T,
    questions: Question[],
    docs?: {
      usageText?: string,
      manPageInfo?: ManPageInfo
    }
  ): Promise<T> {
    const obj: any = { ...argv };

    // Assuming questions is an array of Question objects and argv is the parsed command line arguments object.
    // @ts-ignore
    if (this.noTty && questions.some(question => question.required && this.isEmptyAnswer(argv[question.name]))) {

      for (let index = 0; index < questions.length; index++) {
        const question = questions[index];
        // Apply default value if applicable
        if ('default' in question) {
          obj[question.name] = question.default;
        }
      }

      // @ts-ignore
      if (!questions.some(question => question.required && this.isEmptyAnswer(argv[question.name]))) {
        return obj as T;
      }

      // If running in a non-interactive mode and any required argument is missing, handle the error.
      if (docs?.usageText) {
        this.log(docs.usageText);
      } else if (docs?.manPageInfo) {
        this.log(this.generateManPage(docs.manPageInfo))
      } else {
        this.log('Missing required arguments. Please provide all required parameters.');
      }
      process.exit(1);

    }

    for (let index = 0; index < questions.length; index++) {
      const question = questions[index];
      const ctx: PromptContext = new PromptContext();

      // Apply default value if applicable
      if ('default' in question && (this.useDefaults || question.useDefault)) {
        obj[question.name] = question.default;
        continue;  // Skip to the next question since the default is applied
      }

      while (ctx.needsInput) {
        obj[question.name] = await this.handleQuestionType(question, ctx);

        if (!this.isValid(question, obj, ctx)) {
          continue;
        }
        // If input passes validation and is not empty, or not required, move to the next question
        ctx.nextQuestion();
      }
    }

    return obj as T;
  }

  private async handleQuestionType(question: Question, ctx: PromptContext): Promise<any> {
    switch (question.type) {
      case 'confirm':
        return this.confirm(question as ConfirmQuestion, ctx);
      case 'checkbox':
        return this.checkbox(question as CheckboxQuestion, ctx);
      case 'autocomplete':
        return this.autocomplete(question as AutocompleteQuestion, ctx);
      case 'text':
      default:
        return this.text(question as TextQuestion, ctx);
    }
  }
  async confirm(question: ConfirmQuestion, ctx: PromptContext): Promise<boolean> {
    if (this.noTty || !this.rl) return question.default ?? false;  // Return default if non-interactive

    return new Promise<boolean>((resolve) => {
      this.clearScreen();
      this.rl.question(this.getPrompt(question, ctx, ''), (answer) => {
        const userInput = answer.trim().toLowerCase();

        if (userInput === '') {
          resolve(question.default ?? false);  // Use default value if input is empty
        } else if (['yes', 'y'].includes(userInput)) {
          resolve(true);
        } else {
          resolve(false);
        }
      });
    });
  }

  async text(question: TextQuestion, ctx: PromptContext): Promise<string | null> {
    if (this.noTty || !this.rl) {
      if ('default' in question) {
        return question.default;
      }
      return;
    }

    let input = '';

    return new Promise<string | null>((resolve) => {
      this.clearScreen();
      this.rl.question(this.getPrompt(question, ctx, input), (answer) => {  // Include the prompt directly in the question method
        input = answer;
        if (input.trim() !== '') {
          resolve(input);  // Return input if not empty
        } else {
          resolve(null);  // Return null if empty and not required
        }
      });
    });
  }

  async checkbox(question: CheckboxQuestion, ctx: PromptContext): Promise<OptionValue[]> {
    if (this.noTty || !this.rl) return question.default ?? [];  // Return default if non-interactive

    this.keypress.resume();
    const options = this.sanitizeOptions(question);
    let input = ''; // Search input
    let filteredOptions = options;
    let selectedIndex = 0;
    let startIndex = 0; // Start index for visible options
    const maxLines = this.getMaxLines(question, options.length) // Use provided max or total options
    const selections: boolean[] = new Array(options.length).fill(false);

    const updateFilteredOptions = (): void => {
      filteredOptions = this.filterOptions(options, input);
    };

    const display = (): void => {
      this.clearScreen();
      this.displayPrompt(question, ctx, input);
      const endIndex = Math.min(startIndex + maxLines, filteredOptions.length);
      for (let i = startIndex; i < endIndex; i++) {
        const option = filteredOptions[i];
        const isSelected = selectedIndex === i;
        const marker = isSelected ? '>' : ' ';
        const index = options.map(o => o.name).indexOf(option.name);
        if (index >= 0) {
          const isChecked = selections[index] ? '◉' : '○'; // Use the original index in options
          const line = `${marker} ${isChecked} ${option.name}`;
          this.log(isSelected ? chalk.blue(line) : line);
        } else {
          this.log('No options'); // sometimes user searches and there are no options...
        }
      }
    };

    display();

    // Handling BACKSPACE key
    this.keypress.on(KEY_CODES.BACKSPACE, () => {
      input = input.slice(0, -1);
      updateFilteredOptions();
      display();
    });

    // Register alphanumeric keypresses to accumulate input, excluding space
    'abcdefghijklmnopqrstuvwxyz0123456789'.split('').forEach(char => {
      this.keypress.on(char, () => {
        input += char;
        updateFilteredOptions();
        display();
      });
    });

    this.keypress.on(KEY_CODES.UP_ARROW, () => {
      selectedIndex = selectedIndex > 0 ? selectedIndex - 1 : filteredOptions.length - 1;
      if (selectedIndex < startIndex) {
        startIndex = selectedIndex; // Scroll up
      } else if (selectedIndex === filteredOptions.length - 1) {
        startIndex = Math.max(0, filteredOptions.length - maxLines); // Jump to the bottom of the list
      }
      display();
    });

    this.keypress.on(KEY_CODES.DOWN_ARROW, () => {
      selectedIndex = (selectedIndex + 1) % filteredOptions.length;
      if (selectedIndex >= startIndex + maxLines) {
        startIndex = selectedIndex - maxLines + 1; // Scroll down
      } else if (selectedIndex === 0) {
        startIndex = 0; // Jump to the top of the list
      }
      display();
    });

    this.keypress.on(KEY_CODES.SPACE, () => {
      // Map filtered index back to the original index in options
      selections[options.indexOf(filteredOptions[selectedIndex])] = !selections[options.indexOf(filteredOptions[selectedIndex])];
      display();
    });

    return new Promise<Value[]>(resolve => {
      this.keypress.on(KEY_CODES.ENTER, () => {
        this.keypress.pause();
        const result: Value[] = [];
        if (question.returnFullResults) {
          // Return all options with their selected status
          options.forEach((option, index) => {
            result.push({
              name: option.name,
              value: selections[index]
            });
          });
        } else {
          // Return only options that are selected
          options.forEach((option, index) => {
            if (selections[index]) {
              result.push({
                name: option.name,
                value: selections[index]
              });
            }
          });
        }
        resolve(result);
      });
    });
  }

  async autocomplete(question: AutocompleteQuestion, ctx: PromptContext): Promise<any> {
    if (this.noTty || !this.rl) {
      if ('default' in question) {
        return question.default;
      }
      return;
    }

    this.keypress.resume();
    const options = this.sanitizeOptions(question);

    let input = '';
    let filteredOptions = options;
    let selectedIndex = 0;
    let startIndex = 0;  // Start index for visible options
    const maxLines = this.getMaxLines(question, options.length) // Use provided max or total options

    const display = (): void => {
      this.clearScreen();
      this.displayPrompt(question, ctx, input);
      // Determine the range of options to display
      const endIndex = Math.min(startIndex + maxLines, filteredOptions.length);
      for (let i = startIndex; i < endIndex; i++) {
        const option = filteredOptions[i];
        if (!option) {
          this.log('No options'); // sometimes user searches and there are no options...
        } else if (i === selectedIndex) {
          this.log(chalk.blue('> ' + option.name)); // Highlight the selected option with chalk
        } else {
          this.log('  ' + option.name);
        }
      }
    };

    const updateFilteredOptions = (): void => {
      filteredOptions = this.filterOptions(options, input);
      // Adjust startIndex to keep the selectedIndex in the visible range
      if (selectedIndex < startIndex) {
        startIndex = selectedIndex;
      } else if (selectedIndex >= startIndex + maxLines) {
        startIndex = selectedIndex - maxLines + 1;
      }
      if (selectedIndex >= filteredOptions.length) {
        selectedIndex = Math.max(filteredOptions.length - 1, 0);
      }
    };

    display();

    // Handling BACKSPACE key
    this.keypress.on(KEY_CODES.BACKSPACE, () => {
      input = input.slice(0, -1);
      updateFilteredOptions();
      display();
    });

    // Register alphanumeric and space keypresses to accumulate input
    'abcdefghijklmnopqrstuvwxyz0123456789 '.split('').forEach(char => {
      this.keypress.on(char, () => {
        input += char;
        updateFilteredOptions();
        display();
      });
    });

    // Navigation
    this.keypress.on(KEY_CODES.UP_ARROW, () => {
      selectedIndex = selectedIndex - 1 >= 0 ? selectedIndex - 1 : filteredOptions.length - 1;
      if (selectedIndex < startIndex) {
        startIndex = selectedIndex;  // Scroll up
      } else if (selectedIndex === filteredOptions.length - 1) {
        startIndex = Math.max(0, filteredOptions.length - maxLines); // Jump to the bottom of the list
      }
      display();
    });
    this.keypress.on(KEY_CODES.DOWN_ARROW, () => {
      selectedIndex = (selectedIndex + 1) % filteredOptions.length;
      if (selectedIndex >= startIndex + maxLines) {
        startIndex = selectedIndex - maxLines + 1;  // Scroll down
      } else if (selectedIndex === 0) {
        startIndex = 0;  // Jump to the top of the list
      }
      display();
    });

    return new Promise<OptionValue>(resolve => {
      this.keypress.on(KEY_CODES.ENTER, () => {
        this.keypress.pause();
        resolve(filteredOptions[selectedIndex]?.value || input);
      });
    });
  }

  private getOptionValue(option: string | OptionValue): OptionValue {
    if (typeof option === 'string') {
      return { name: option, value: option };
    } else if (typeof option === 'object' && option && 'name' in option) {
      return { name: option.name, value: option.value };
    } else {
      return undefined;
    }
  }

  private sanitizeOptions(question: AutocompleteQuestion | CheckboxQuestion): OptionValue[] {
    const options = (question.options ?? []).map(option => this.getOptionValue(option));
    return options.filter(Boolean);
  }

  filterOptions(options: OptionValue[], input: string): OptionValue[] {
    input = input.toLowerCase(); // Normalize input for case-insensitive comparison

    // Fuzzy matching: Check if all characters of the input can be found in the option name in order
    const fuzzyMatch = (option: string, input: string) => {
      const length = input.length;
      let position = 0; // Position in the input string

      // Iterate over each character in the option name
      for (let i = 0; i < option.length; i++) {
        if (option[i] === input[position]) {
          position++; // Move to the next character in the input
          if (position === length) { // Check if we've matched all characters
            return true;
          }
        }
      }
      return false;
    };

    return options
      .filter(option => fuzzyMatch(option.name.toLowerCase(), input))
      .sort((a, b) => {
        if (a.name < b.name) {
          return -1;
        }
        if (a.name > b.name) {
          return 1;
        }
        return 0;
      });
  }

  getMaxLines(question: { maxDisplayLines?: number }, defaultLength: number): number {
    if (question.maxDisplayLines) {
      return question.maxDisplayLines;
    }

    // if (!this.noTty && (this.output as any).isTTY) {
    //   const rows = Math.round(((this.output as any).rows ?? 0) / 7);
    //   return Math.max(rows, defaultLength);
    // }
    return Math.min(this.globalMaxLines, defaultLength);
  }

  // Method to cleanly close the readline interface
  public close() {
    if (this.rl) {
      this.rl.close();
      this.keypress.destroy();
    }
  }
}