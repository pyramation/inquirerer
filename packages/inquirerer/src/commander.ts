import deepmerge from 'deepmerge';
import minimist, { Opts, ParsedArgs } from 'minimist';
import { Readable,Writable } from 'stream';

import { Inquirerer } from './prompt';
import { getVersion } from './utils';

// Define the type for the command handler function
export type CommandHandler = (argv: ParsedArgs, prompter: Inquirerer, options: CLIOptions) => void;

export interface CLIOptions {
  noTty: boolean;
  input: Readable;
  output: Writable;
  minimistOpts: Opts;
  version: string;
}

export const defaultCLIOptions: CLIOptions = {
  version: `inquirerer@${getVersion()}`,
  noTty: false,
  input: process.stdin,
  output: process.stdout,
  minimistOpts: {
    alias: {
      v: 'version'
    }
  }
};

export class CLI {
  private argv: ParsedArgs;
  private prompter: Inquirerer;
  private commandHandler: CommandHandler;
  private options: CLIOptions;

  constructor(
    commandHandler: CommandHandler,
    options: Partial<CLIOptions>,
    argv?: any
  ) {
    const { input, output, ...optionsWithoutIO } = options;
    const { input: defaultInput, output: defaultOutput, ...defaultOptionsWithoutIO } = defaultCLIOptions;
    const mergedOptions: Partial<CLIOptions> = deepmerge(defaultOptionsWithoutIO, optionsWithoutIO);
    mergedOptions.input = input || defaultInput;
    mergedOptions.output = output || defaultOutput;
    this.options = mergedOptions as CLIOptions;

    this.argv = argv ? argv : minimist(process.argv.slice(2), this.options.minimistOpts);

    this.prompter = new Inquirerer({
      noTty: this.options.noTty,
      input: this.options.input,
      output: this.options.output
    });
    this.commandHandler = commandHandler;

  }

  public async run(): Promise<void> {
    if (!('tty' in this.argv)) {
      this.argv.tty = true;
    }

    if (this.argv.version) {
      console.log(this.options.version);
      process.exit(0);
    }

    const args = await this.commandHandler(this.argv, this.prompter, this.options);
    this.prompter.close();
    return args;
  }
}

export default CLI;
