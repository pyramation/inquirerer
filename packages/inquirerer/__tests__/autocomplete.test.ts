import readline from 'readline';
import { Readable, Writable, Transform } from 'stream';

import stripAnsi from 'strip-ansi';
import { Inquirerer } from '../src';
import { Question } from '../src/question';

jest.mock('readline');

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

const snap = (str: any) => expect(str).toMatchSnapshot();

describe('Inquirerer', () => {
  let mockWrite: jest.Mock;
  let mockInput: Readable;
  let mockOutput: Writable;
  let questionHandlers: Array<(input: string) => void> = [];
  let currentQuestionIndex: number = 0;
  let transformStream: Transform;

  let writeResults: string[];
  let transformResults: string[];

  let inputQueue: Array<{ type: 'key' | 'read', value: string }> = [];
  let currentInputIndex: number = 0;

  function setupReadlineMock() {
    readline.createInterface = jest.fn().mockReturnValue({
      question: (questionText: string, cb: (input: string) => void) => {
        // Process the queued inputs when question is called
        const nextInput = inputQueue[currentInputIndex++];
        if (nextInput && nextInput.type === 'read') {
          setTimeout(() => cb(nextInput.value), 350); // Simulate readline delay
        }
      },
      close: jest.fn(),
    });
  }

  function enqueueInputResponse(input: { type: 'key' | 'read', value: string }) {
    if (input.type === 'key') {
  // Push key events directly to mockInput
  // @ts-ignore
      setTimeout(() => mockInput.push(input.value), 350);
    } else {
      // Queue readline responses to be handled by the readline mock
      inputQueue.push(input);
    }
  }

  const KEY_SEQUENCES = {
    ENTER: '\u000d',     // Carriage return
    UP_ARROW: '\u001b[A', // ANSI escape sequence for the up arrow
    DOWN_ARROW: '\u001b[B',
    SPACE: ' '
  };

  beforeEach(() => {
    mockWrite = jest.fn();
    currentQuestionIndex = 0;
    questionHandlers = [];
    writeResults = [];
    transformResults = [];


    mockInput = new Readable({
      read(size) { }
    });
    // @ts-ignore
    mockInput.setRawMode = jest.fn();  // Mock TTY-specific method if needed

    mockOutput = new Writable({
      write: (chunk, encoding, callback) => {
        const str = chunk.toString();
        writeResults.push(stripAnsi(str));
        mockWrite(str);
        callback();
      }
    });

    // Create the transform stream to log and pass through data
    transformStream = new Transform({
      transform(chunk, encoding, callback) {
        const data = chunk.toString();
        transformResults.push(stripAnsi(data));
        this.push(chunk); // Pass the data through
        callback();
      }
    });

    setupReadlineMock();
    // Pipe the transform stream to the mock output to intercept writes
    //  transformStream.pipe(mockOutput);

    // mockOutput.pipe(transformStream);
    mockInput.pipe(transformStream);

    // mockInput.pipe(transformStream).pipe(mockOutput);

  });

  it('handles autocomplete selection with key events', async () => {
    const prompter = new Inquirerer(false, mockInput, mockOutput);
    const questions: Question[] = [{
      name: 'autocompleteField',
      type: 'autocomplete',
      options: ['first option', 'firry second option', 'firry third option']
    }];
    const initialParams = {};

    enqueueInputResponse({ type: 'read', value: 'fir' });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.DOWN_ARROW });
    enqueueInputResponse({ type: 'key', value: KEY_SEQUENCES.ENTER });

    const result = await prompter.prompt(initialParams, questions);

    // Expected to select the second option
    // expect(result).toEqual({ autocompleteField: 'second option' });
    snap({result});
    snap(writeResults);
    snap(transformResults);
  });


});
