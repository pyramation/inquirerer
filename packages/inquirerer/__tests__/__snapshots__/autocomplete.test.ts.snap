// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inquirerer handles autocomplete selection with key events 1`] = `
{
  "result": {
    "autocompleteField": "firry third option",
  },
}
`;

exports[`Inquirerer handles autocomplete selection with key events 2`] = `
[
  "c",
  "autocompleteField:
> 
",
  "[34m> first option[39m
",
  "  firry second option
",
  "  firry third option
",
  "c",
  "autocompleteField:
> 
",
  "  first option
",
  "[34m> firry second option[39m
",
  "  firry third option
",
  "c",
  "autocompleteField:
> 
",
  "  first option
",
  "  firry second option
",
  "[34m> firry third option[39m
",
]
`;

exports[`Inquirerer handles autocomplete selection with key events 3`] = `
[
  "[B",
  "[B",
  "
",
]
`;
